File: Lokesh_Saana/131-palindrome-partitioning.py
class Solution:

    def partition(self, s: str) -> List[List[str]]:
        if len(s) == 0:
            return []
        if len(s) == 1:
            return [s]

        def ispal(s: str) -> bool:
            # return s == s[::-1]
            l, r = 0, len(s)-1
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True

        ans = [] if not ispal(s) else [[s]]
        for size in range(1, len(s)):
            if ispal(s[:size]):
                for rest in self.partition(s[size:]):
                    ans.append([s[:size], *rest])
        return ans

File: Ayush-Gupta/131. Palindrome Partitioning.cpp
class Solution {
public:
    bool isPal(string &s, int i, int j) {
        while(i<j) {
            if(s[i++] != s[j--]) return 0;
        }
        return 1;
    }
    void f(int ind, string &s, vector<string> &ds, vector<vector<string>> &ans) {
        if(ind==s.size()) {
            ans.push_back(ds);
            return;
        }
        for(int i=ind;i<s.size();i++) {
            if(isPal(s, ind, i)) {
                ds.push_back(s.substr(ind, i-ind+1));
                f(i+1, s, ds, ans);
                ds.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> ds;
        f(0, s, ds, ans);
        return ans;
    }
};